ReactAndroid/src/main/java/com/facebook/react/uimanager/animation/NativeAnimatedNodesManager.java

```diff
@@ -0,0 +1,344 @@
+    // step). We store number of visited nodes in this step in `updatedNodesCount`
+
+    nodesQueue.clear();
+    mAnimatedGraphDFSColor++;
+    if (mAnimatedGraphDFSColor == AnimatedNode.INITIAL_DFS_COLOR) {
+      // see reasoning for this check a few lines above
+      mAnimatedGraphDFSColor++;
+    }
+
+    // find nodes with zero "incoming nodes", those can be either nodes from `mUpdatedNodes` or
+    // ones connected to active animations
+    for (int i = 0; i < mUpdatedNodes.size(); i++) {
+      AnimatedNode node = mUpdatedNodes.get(i);
+      if (node.mActiveIncomingNodes == 0 && node.mDFSColor != mAnimatedGraphDFSColor) {
+        node.mDFSColor = mAnimatedGraphDFSColor;
+        updatedNodesCount++;
+        nodesQueue.add(node);
+      }
+    }
+    for (int i = 0; i < mActiveAnimations.size(); i++) {
+      AnimationDriver animation = mActiveAnimations.get(i);
+      AnimatedNode valueNode = animation.mAnimatedValue;
+      if (valueNode.mActiveIncomingNodes == 0 && valueNode.mDFSColor != mAnimatedGraphDFSColor) {
+        valueNode.mDFSColor = mAnimatedGraphDFSColor;
+        updatedNodesCount++;
+        nodesQueue.add(valueNode);
+      }
+    }
+
+    // Run main "update" loop
+    ArrayList<PropsAnimatedNode> updatedPropNodes = new ArrayList<>();
+    while (!nodesQueue.isEmpty()) {
+      AnimatedNode nextNode = nodesQueue.poll();
+      nextNode.runAnimationStep(frameTimeNanos);
+      if (nextNode instanceof PropsAnimatedNode) {
+        updatedPropNodes.add((PropsAnimatedNode) nextNode);
+      }
+      if (nextNode.mChildren != null) {
+        for (int i = 0; i < nextNode.mChildren.size(); i++) {
+          AnimatedNode child = nextNode.mChildren.get(i);
+          child.feedDataFromUpdatedParent(nextNode);
+          child.mActiveIncomingNodes--;
+          if (child.mDFSColor != mAnimatedGraphDFSColor && child.mActiveIncomingNodes == 0) {
+            child.mDFSColor = mAnimatedGraphDFSColor;
+            updatedNodesCount++;
+            nodesQueue.add(child);
+          }
+        }
+      }
+    }
+
+    // Verify that we've visited *all* active nodes. Throw otherwise as this would mean there is a
+    // cycle in animated node graph. We also take advantage of the fact that all active nodes are
+    // visited in the step above so that all the nodes properties `mActiveIncomingNodes` are set to
+    // zero
+    if (activeNodesCount != updatedNodesCount) {
+      throw new IllegalStateException("Looks like animated nodes graph has cycles, there are "
+        + activeNodesCount + " but toposort visited only " + updatedNodesCount);
+    }
+
+    // Collect UI updates
+    for (int i = 0; i < updatedPropNodes.size(); i++) {
+      PropsAnimatedNode propNode = updatedPropNodes.get(i);
+      PropsAnimatedNode.UpdateViewData data = propNode.createUpdateViewData();
+      if (data.mViewTag > 0) {
+        mEnqueuedUpdates.add(propNode.createUpdateViewData());
```

Comment:
That should be fine. I'll update that. I used to run this code on "native thread". That's why I needed a queue here, but it's no longer necessary